### ISSUE TITLE
Syslog Unable to Load With Large Programs Table (Plus a Solution)

### ISSUE THREAD

**Issue by stueh** [Cacti/plugin_syslog#30]
Hi All,

tl;dr Probably need to add a query to cleanup orphaned entries in the syslog_programs table after running removal/delete rules. The following would work:

`DELETE FROM syslog_programs where program_id not in (Select program_id from syslog) OR (Select program_id from syslog_removed)`

----------------------------------------------------------

I've had an issue for a while now where after some time of running without a problem, Syslog (the page) will take longer and longer to load, until it finally fails to load entirely - it'll simply display a blank page for several minutes until timeout or if using Chrome, the "Wait/Kill" window appears.

Now I've eventually narrowed it down to the syslog_programs table being quite full - when the table is emptied Syslog loads fine again without an issue, so it is an issue with the syslog_programs table having many entries (e.g. 90,000+ which slows down loading). Emptying the table is a bad idea, though, as syslog won't appear if there's no program_id tying it to syslog_programs.

What's interesting here is that those programs defined in syslog_programs are essentially orphaned - CentOS (this is what our Cacti is running on) is providing a different program ID every time cron runs polling, and I have a removal rule in place to remove any syslog entries created by the host (as these are in the Cacti Log tab). It looks like when Syslog processes the incoming messages, it creates the program entries in the syslog_programs table, and THEN deletes the syslog entry, leaving the entries in the syslog_program tables orphaned and unneeded - when it fills up, syslog slows down and we're stuck in the mud.

I found a temporary solution which I can run by hand, but it may be a good idea to incorporate this into the script for removals? I don't know enough about PHP programming to modify it myself!

So, take a look at the size of the syslog_programs table [in this screenshot here](http://i.imgur.com/3yGfSYu.jpg) and you'll see I have 27,740 entries.

I can find the "orphaned" syslog entries by using the SQL query:
`SELECT * FROM syslog_programs where program_id not in (Select program_id from syslog)`
[See this picture here](http://i.imgur.com/Td9iQdp.jpg)

I then delete the "orphaned" entries by using the query:
`DELETE FROM syslog_programs where program_id not in (Select program_id from syslog)`
[See picture here](http://i.imgur.com/JS4gx20.jpg)
NB: In hindsite should probably use the following in case there are any rules which are for removal instead of deletion: `DELETE FROM syslog_programs where program_id not in (Select program_id from syslog) OR (Select program_id from syslog_removed)`

And now the table is back down to a reasonable size of 10 programs, [see here](http://i.imgur.com/EwAa5fq.jpg)

At the moment I'm running this query by hand when syslog slows down enough, but I'm thinking this should maybe be incorporated into the script for syslog removals?

Just and idea for a solution to my problems :)

**Comment by cigamit** [comment]
Thanks for the tip/hint.  Need to do two thing, purge the programs as you indicated, and make the programs dropdown an autocomplete.  Working on it.

**Comment by cigamit** [comment]
Resolved.
